src/P256Verifier.sol

```diff
@@ -58,286 +58,189 @@ contract P256Verifier {
-
-                                Y := addmod(T2, mulmod(T1, Y, p), p) //Y3= M(S-X3)-W*Y1
-
-                                continue
-                            }
-                        }
-
-                        T4 := mulmod(T2, T2, p) //PP
-                        let TT1 := mulmod(T4, T2, p) //PPP, this one could be spared, but adding this register spare gas
-                        zz := mulmod(zz, T4, p)
-                        zzz := mulmod(zzz, TT1, p) //zz3=V*ZZ1
-                        let TT2 := mulmod(X, T4, p)
-                        T4 := addmod(
-                            addmod(mulmod(y2, y2, p), sub(p, TT1), p),
-                            mulmod(minus_2, TT2, p),
-                            p
-                        )
-                        Y := addmod(
-                            mulmod(addmod(TT2, sub(p, T4), p), y2, p),
-                            mulmod(Y, TT1, p),
-                            p
-                        )
-
-                        X := T4
-                    }
-                } //end loop
-                mstore(add(T, 0x60), zz)
-                //(X,Y)=ecZZ_SetAff(X,Y,zz, zzz);
-                //T[0] = inverseModp_Hard(T[0], p); //1/zzz, inline modular inversion using precompile:
-                // Define length of base, exponent and modulus. 0x20 == 32 bytes
-                mstore(T, 0x20)
-                mstore(add(T, 0x20), 0x20)
-                mstore(add(T, 0x40), 0x20)
-                // Define variables base, exponent and modulus
-                //mstore(add(pointer, 0x60), u)
-                mstore(add(T, 0x80), minus_2)
-                mstore(add(T, 0xa0), p)
-
-                // Call the precompiled contract 0x05 = ModExp
-                if iszero(call(not(0), 0x05, 0, T, 0xc0, T, 0x20)) {
-                    revert(0, 0)
-                }
-
-                //Y:=mulmod(Y,zzz,p)//Y/zzz
-                //zz :=mulmod(zz, mload(T),p) //1/z
-                //zz:= mulmod(zz,zz,p) //1/zz
-                X := mulmod(X, mload(T), p) //X/zz
-            } //end assembly
-        } //end unchecked
-
-        return X;
+        if (scalar_u == 0 && scalar_v == 0) return (0, true);
+
+        (H0, H1, add_success) = ecAff_add(gx, gy, Q0, Q1);
+
+        if (!add_success) {
+            return (0, false);
+        }
+
+        int256 index = 255;
+        uint256 bitpair;
+
+        // Find the first bit index that's active in either scalar_u or scalar_v.
+        while(index >= 0) {
+            bitpair = compute_bitpair(uint256(index), scalar_u, scalar_v);
+            index--;
+            if (bitpair != 0) break;
+        }
+
+        // initialise (X, Y) depending on the first active bitpair.
+        if (bitpair == 0) { // if scalar_u = scalar_v = 0.
+            X = 0;
+            Y = 0;
+        }
+        else if (bitpair == 1) {
+            X = gx;
+            Y = gy;
+        } else if (bitpair == 2) {
+            X = Q0;
+            Y = Q1;
```

Comment:
might be worth calling this `(QX, QY)` or conversely `g0, g1`

Reply:
done